vm/systemSmartContracts/delegation.go

```diff
@@ -605,16 +864,189 @@ func (d *delegation) addValueToFund(key []byte, value *big.Int) error {
 	return err
 }
 
-func (d *delegation) unDelegate(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+func (d *delegation) unDelegate(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	err := d.eei.UseGas(d.gasCost.MetaChainSystemSCsCost.DelegationOps)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.OutOfGas
+	}
+	if len(args.Arguments) != 1 {
+		d.eei.AddReturnMessage("wrong number of arguments")
+		return vmcommon.FunctionWrongSignature
+	}
+	valueToUnDelegate := big.NewInt(0).SetBytes(args.Arguments[0])
+	if valueToUnDelegate.Cmp(zero) <= 0 {
+		d.eei.AddReturnMessage("invalid value to undelegate")
+		return vmcommon.UserError
+	}
+
+	isNew, delegator, err := d.getOrCreateDelegatorData(args.CallerAddr)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if isNew {
+		d.eei.AddReturnMessage("it is not a delegator")
+		return vmcommon.UserError
+	}
+
+	activeFund, err := d.getFund(delegator.ActiveFund)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if activeFund.Value.Cmp(valueToUnDelegate) < 0 {
+		d.eei.AddReturnMessage("invalid value to undelegate")
+		return vmcommon.UserError
+	}
+
+	activeFund.Value.Sub(activeFund.Value, valueToUnDelegate)
+	err = d.saveFund(delegator.ActiveFund, activeFund)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	unStakedFundKey, err := d.createAndSaveNextFund(args.CallerAddr, valueToUnDelegate, unStaked)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	globalFund, err := d.getGlobalFundData()
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	globalFund.UnStakedFunds = append(globalFund.UnStakedFunds, unStakedFundKey)
+	globalFund.TotalActive.Sub(globalFund.TotalActive, valueToUnDelegate)
```

Comment:
we should not allow subtracting TotalActive under the stakedNodes * nodePrice 
We have to think how we can trigger a node unstake event (that can not be reverted by the delegation owner) in such situation

Reply:
this will come in another PR. the funds protection part is not resolved.

Reply:
ok